cryptconverter/data-converter.go

```diff
@@ -1,94 +1,95 @@
 	MetadataEncryptionKeyId = "encryption-key-id"
 
-	// MetadataContentEncoding is "content-encoding"
-	MetadataContentEncoding = "content-encoding"
+	// MetadataEncodingEncrypted is "binary/encrypted"
+	MetadataEncodingEncrypted = "binary/encrypted"
 )
 
+// Work around ContextAware being internal to SDK
+type ContextAware interface {
+	WithWorkflowContext(ctx workflow.Context) converter.DataConverter
+	WithContext(ctx context.Context) converter.DataConverter
+}
+
 // CryptDataConverter implements DataConverter using AES Crypt.
 type CryptDataConverter struct {
 	dataConverter converter.DataConverter
+	context       CryptContext
 }
 
-// getEncryptionKey fetches the crypt key from secure storage
-func (dc *CryptDataConverter) getEncryptionKey() (keyId string, key []byte) {
-	// Key can be fetched from KMS or other secure storage.
-	return "test", []byte("test-key-test-key-test-key-test!")
-}
-
-// getDecryptionKey fetches the crypt key from secure storage
-func (dc *CryptDataConverter) getDecryptionKey(keyId string) (key []byte) {
-	// Key can be fetched from KMS or other secure storage.
-	return []byte("test-key-test-key-test-key-test!")
-}
+func (dc *CryptDataConverter) WithWorkflowContext(ctx workflow.Context) converter.DataConverter {
+	if val := ctx.Value(PropagateKey); val != nil {
+		dataConverter := dc.dataConverter
+		if dcwc, ok := dc.dataConverter.(ContextAware); ok {
+			dataConverter = dcwc.WithWorkflowContext(ctx)
+		}
 
-// NewCryptDataConverter created new instance of CryptDataConverter wrapping a DataConverter
-func NewCryptDataConverter(dataConverter converter.DataConverter) *CryptDataConverter {
-	return &CryptDataConverter{
-		dataConverter: dataConverter,
+		return &CryptDataConverter{
+			dataConverter: dataConverter,
+			context:       val.(CryptContext),
+		}
 	}
-}
 
-func encrypt(plainData []byte, key []byte) ([]byte, error) {
-	c, err := aes.NewCipher(key)
-	if err != nil {
-		return nil, err
-	}
+	return dc
+}
 
-	gcm, err := cipher.NewGCM(c)
-	if err != nil {
-		return nil, err
-	}
+func (dc *CryptDataConverter) WithContext(ctx context.Context) converter.DataConverter {
+	if val := ctx.Value(PropagateKey); val != nil {
+		dataConverter := dc.dataConverter
+		if dcwc, ok := dc.dataConverter.(ContextAware); ok {
+			dataConverter = dcwc.WithContext(ctx)
+		}
 
-	nonce := make([]byte, gcm.NonceSize())
-	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
-		return nil, err
+		return &CryptDataConverter{
+			dataConverter: dataConverter,
+			context:       val.(CryptContext),
+		}
 	}
 
-	return gcm.Seal(nonce, nonce, plainData, nil), nil
+	return dc
 }
 
-func decrypt(encryptedData []byte, key []byte) ([]byte, error) {
-	c, err := aes.NewCipher(key)
-	if err != nil {
-		return nil, err
-	}
-
-	gcm, err := cipher.NewGCM(c)
-	if err != nil {
-		return nil, err
-	}
+func (dc *CryptDataConverter) getKey(keyId string) (key []byte) {
+	// Key can be fetched from KMS or other secure storage.
```

Comment:
nit: I'd probably make this wording stronger. Maybe even "*must* be fetched ... in a production setting"